include/mbedtls/mps/trace.h

```diff
@@ -0,0 +1,113 @@
+
+#ifndef MBEDTLS_MPS_TRACE_H
+#define MBEDTLS_MPS_TRACE_H
+
+/*
+ * Adapt this to enable/disable tracing output
+ * from the various layers of the MPS.
+ */
+
+#define TRACE_ENABLE_LAYER_1
+#define TRACE_ENABLE_LAYER_2
+#define TRACE_ENABLE_LAYER_3
+#define TRACE_ENABLE_READER
+#define TRACE_ENABLE_WRITER
+
+/*
+ * This shouldn't be touched.
+ */
+
+__attribute__((unused)) static int trace_id;
+
+typedef enum
+{
+    trace_comment=0,
+    trace_call,
+    trace_error
+} trace_type;
+
+#define TRACE_ID_LAYER_1 1
+#define TRACE_ID_LAYER_2 2
+#define TRACE_ID_LAYER_3 3
+#define TRACE_ID_WRITER  4
+#define TRACE_ID_READER  5
+
+#if defined(TRACE_ENABLE_LAYER_1)
+#define TRACE_MASK_LAYER_1 (1u << TRACE_ID_LAYER_1 )
+#else
+#define TRACE_MASK_LAYER_1 0
+#endif
+
+#if defined(TRACE_ENABLE_LAYER_2)
+#define TRACE_MASK_LAYER_2 (1u << TRACE_ID_LAYER_2 )
+#else
+#define TRACE_MASK_LAYER_2 0
+#endif
+
+#if defined(TRACE_ENABLE_LAYER_3)
+#define TRACE_MASK_LAYER_3 (1u << TRACE_ID_LAYER_3 )
+#else
+#define TRACE_MASK_LAYER_3 0
+#endif
+
+#if defined(TRACE_ENABLE_READER)
+#define TRACE_MASK_READER (1u << TRACE_ID_READER )
+#else
+#define TRACE_MASK_READER 0
+#endif
+
+#if defined(TRACE_ENABLE_WRITER)
+#define TRACE_MASK_WRITER (1u << TRACE_ID_WRITER )
+#else
+#define TRACE_MASK_WRITER 0
+#endif
+
+#define TRACE_MASK ( TRACE_MASK_LAYER_1 |           \
+                     TRACE_MASK_LAYER_2 |           \
+                     TRACE_MASK_LAYER_3 |           \
+                     TRACE_MASK_READER  |           \
+                     TRACE_MASK_WRITER )
+
+/* We have to avoid globals because E-ACSL chokes on them...
+ * Wrap everything in stub functions. */
+int get_trace_depth( void );
+void inc_trace_depth( void );
+void dec_trace_depth( void );
+
+void trace_color( int id );
+void trace_indent( int level, trace_type ty );
+
+#define TRACE( type, fmt, ... )                                         \
+    do {                                                                \
+        if( ! ( TRACE_MASK & ( 1u << trace_id ) ) )                     \
```

Comment:
I agree that there might be a less clumsy way to implement this, but the compiler should be smart enough to eliminate these sections at compile time. Given that there are more urgent things to address, I'd prefer to keep this as is for now - ok?

Reply:
Compilers should be able to eliminate the runtime branching, so I don't think this is a critical issue. On the other hand, if tracing happens often enough, this may still impact debug builds. This may result in increased running times for the developers or in some CI jobs, so it may still be worth fixing in the future. This may create a false impression, that MPS is slow (I have not actually measured this; it may as well be negligible). Also, this doesn't seem to be a necessary cost. Maybe an issue or a task can be recorded somewhere, so that we don't forget about it. A profiling/optimizing task may be planned for the later stages of MPS deployment; it can be addressed properly (i.e. with measurement) then.